### PR TITLE
Added file validation to iconv

### PR DESCRIPTION
**Pull request by tom-wa**

### CODE REVIEW

File: src/plugins/iconv/iconv.c [hunk 1]
```diff
@@ -93,12 +106,12 @@ setlocale (LC_ALL, "");
  * this plugin can't be used.
  *
  * @param direction must be @c UTF8_TO (convert from current non-UTF-8 to
- * 	UTF-8) or @c UTF8_FROM (convert from UTF-8 to current non-UTF-8)
```
Comment (markus2330) [inline src/plugins/iconv/iconv.c]: spaces?

File: src/plugins/iconv/README.md [hunk 2]
```diff
@@ -3,12 +3,13 @@
...
 - infos/provides = conv
-- infos/placements = postgetstorage presetstorage
+- infos/placements = pregetstorage postgetstorage presetstorage
 - infos/description = Converts values of keys between charsets
 
 ## Introduction ##
 
-This plugin is a filter plugin that converts between different character encodings.
+This plugin is a filter plugin that converts between different character encodings,
+or, if the `checkfile` key is set, validates the fileencoding.
```
Comment (markus2330) [inline src/plugins/iconv/README.md]: only validates or also transforms?

File: src/plugins/iconv/iconv.c [hunk 3]
```diff
@@ -158,6 +170,40 @@ int kdbbUTF8Engine(Plugin *handle, int direction, char **string, size_t *inputOu
 	return 0;
 }
 
+static int validate_file(Plugin *handle, Key *parentKey)
```
Comment (markus2330) [inline src/plugins/iconv/iconv.c]: please use camelCase

File: src/plugins/iconv/iconv.c [hunk 4]
```diff
@@ -26,6 +26,11 @@
 
 
 #include "iconv.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <errno.h>
+
+#define CHAR_SIZE_MAX 4
```
Comment (markus2330) [inline src/plugins/iconv/iconv.c]: Why only iconv with size of 4?

Comment (tom-wa) [inline src/plugins/iconv/iconv.c]: the wides character encoding i know is 4 bytes long. the OS should deal with the rest of the buffering.

File: src/plugins/iconv/iconv.c (hunk 1 above)
Comment (tom-wa) [inline src/plugins/iconv/iconv.c]: vim messed up...

File: src/plugins/iconv/iconv.c (hunk 4 above)
Comment (markus2330) [inline src/plugins/iconv/iconv.c]: The older standard RFC 2279 seems to also know 6 bytes. But thats not the point, it seems to be very inefficient to handle every character seperately. Any specific reason why kdbbUTF8Engine is not used?

If the kdbbUTF8Engine implementation is not good, it should be fixed.

**Comment by tom-wa** [comment]
@markus2330 done, can be merged now. had to change the return value of kdbbUTF8Engine

File: src/plugins/iconv/README.md [hunk 5]
```diff
@@ -3,12 +3,13 @@
...
 - infos/provides = conv
-- infos/placements = postgetstorage presetstorage
+- infos/placements = pregetstorage postgetstorage presetstorage
 - infos/description = Converts values of keys between charsets
 
 ## Introduction ##
 
-This plugin is a filter plugin that converts between different character encodings.
+This plugin is a filter plugin that converts between different character encodings,
+or, if the `checkfile` key is set, validates the fileencoding before reading the file.
```
Comment (markus2330) [inline src/plugins/iconv/README.md]: `checkfile` configuration key..

Please give example how to use.

File: src/plugins/iconv/iconv.c [hunk 6]
```diff
@@ -95,12 +109,12 @@ setlocale (LC_ALL, "");
  * @param direction must be @c UTF8_TO (convert from current non-UTF-8 to
  * 	UTF-8) or @c UTF8_FROM (convert from UTF-8 to current non-UTF-8)
  * @param string before the call: the string to be converted; after the call:
- * 	reallocated to carry the converted string
```
Comment (markus2330) [inline src/plugins/iconv/iconv.c]: still issues with spaces

File: src/plugins/iconv/iconv.c [hunk 7]
```diff
@@ -117,15 +131,25 @@ int kdbbUTF8Engine(Plugin *handle, int direction, char **string, size_t *inputOu
...
+	
+	int Multiplier;
+	
+	if(kdbbNeedsUTF8Conversion(handle))
+	{
+		Multiplier = 4;
+	}
+	else
+	{
+    	Multiplier = 1;
```
Comment (markus2330) [inline src/plugins/iconv/iconv.c]: issues with spaces

File: src/plugins/iconv/iconv.c [hunk 8]
```diff
@@ -184,8 +232,15 @@ int elektraIconvGet(Plugin *handle, KeySet *returned, Key *parentKey)
 		ksDel (pluginConfig);
 		return 1;
 	}
-
-	if (!kdbbNeedsUTF8Conversion(handle)) return 0;
```
Comment (markus2330) [inline src/plugins/iconv/iconv.c]: This should still be present in the case that getFileFlag is false.

File: src/plugins/iconv/iconv.c [hunk 9]
```diff
@@ -232,8 +287,6 @@ int elektraIconvSet(Plugin *handle, KeySet *returned, Key *parentKey)
 	Key *cur;
 	const Key *meta;
 
-	if (!kdbbNeedsUTF8Conversion(handle)) return 0;
```
Comment (markus2330) [inline src/plugins/iconv/iconv.c]: same here

**Comment by markus2330** [comment]
Can you please add a testcase for validation?

File: src/plugins/iconv/iconv.c (hunk 8 above)
Comment (tom-wa) [inline src/plugins/iconv/iconv.c]: the problem is that if you mount iconv with from=UTF-8 and no to= or to=UTF-8 no validation is happening, thats why i removed it.

Comment (markus2330) [inline src/plugins/iconv/iconv.c]: Yes, but your statement is only valid if validation is happening at all.
Iconv is not only for checking, but also for converting.

Honestly, the optimization of from==to is not really important anymore.
If you do not want to use iconv, you would not insert the plugin at all.

But, is it necessary to use iconv the whole text twice for 
validation+transformation? This should be at least documented.

**Comment by tom-wa** [comment]
@markus2330 done

**Comment by markus2330** [comment]
thanks!

**Pull request merged**